php/libraries/NDB_Form.class.inc

```diff
@@ -145,6 +217,49 @@ class NDB_Form extends NDB_Page
      */
     function _save($values)
     {
+
+        // Validate CSRF token, if one is present.  In the future they should be
+        // included with every state-changing request (usually POST requests
+        // but LORIS does not always conform to this).  For now this will check
+        // only that when a CSRF token is sent that it is well-formed and
+        // accurate but will not enforce failure if one is not present.
+        // TODO Once CSRF tokens are enforced universally across LORIS, any
+        // state-changing request wihtout a valid CSRF token should be rejected.
+
+        $token = $values['__csrfToken'] ?? null;
+        if (!is_null($token)) {
+            $authenticRequest = false;
+            $factory          = \NDB_Factory::singleton();
+            $config           = $factory->config();
+            $key = $config->getSetting("JWTKey");
+
+            // Decode token and exit if it is invalid.
+            try {
+                $decoded = \Firebase\JWT\JWT::decode($token, $key, array("HS256"));
+            } catch (Exception $e) {
+                error_log(
+                    "ERROR: Could not decode JWT CSRF token! Possible CSRF
+                    attack. $e"
+                );
+                // Do not do any further processing of the request if given a
+                // bad token.
+                die("Malformed CSRF token. Exiting.");
+            }
+
+            // This is the recommended way to get values from the JWT token
+            // according to the Firebase documentation.
+            $decodedArray = (array) $decoded;
+            $realUser     = $_SESSION['State']->getUsername();
+            error_log($realUser);
+            $authenticRequest = ($decodedArray['user'] === $realUser);
+
+            if (!$authenticRequest) {
+                error_log("ERROR: Bad CSRF token! Possible hacking attempt. $e");
```

Comment:
Why just log the possible hack attempt here when at line 241, the code dies after logging the error?

Reply:
I could be wrong but I think it's because line 239 will only catch the case where the JWT token is invalid, whereas 256 will handle the case of a valid (read: decodable) JWT token that, for whatever reason, had an invalid username.

Reply:
Yeah it's basically to distinguish a "bad" token (not well-formed and so not decode-able by the library" and an "evil" token (which can be decoded but has deliberately false information).

Reply:
Ok thanks!